test/GaslessV3.js

```diff
@@ -304,44 +311,110 @@ function describeTestsForGaslessApproval(data) {
         let tokenAddress = data.tokenAddress;
 
         this.beforeAll(async () => {
-            token = await ethers.getContractAt('ERC20', tokenAddress, owner);
+            token = await ethers.getContractAt('ERC20Upgradeable', tokenAddress, owner);
         });
 
         it('Get token from Uniswap', async () => {
             await getTokenFromUniswap(data, token, tokenAddress);
         });
 
         it('Get approval', async () => {
+            let holder, expiry, allowedDai;
+            // DAI variables
+            holder = owner.address;
+            expiry = Math.round(new Date().getTime() / 1000 + 10_000);
+            allowedDai = true;
+
             let deadline = Math.round(new Date().getTime() / 1000 + 10_000);
             let value = ethers.utils.parseEther('10000').toString();
 
+            console.log('getting tk nonces contract...');
+
             let tokenNonces = await ethers.getContractAt(
                 'ERC20Nonces',
                 tokenAddress
             );
-            let tokenNonce = parseInt(await tokenNonces.nonces(owner.address));
-            let {
-                r: approvalSigR,
-                s: approvalSigS,
-                v: approvalSigV,
-            } = await getSignature({
-                wallet: owner,
-                message: {
-                    owner: owner.address,
-                    spender: main.address,
-                    value: value,
-                    nonce: tokenNonce,
-                    deadline,
-                },
-                messageType: TestCases.constants.permitType,
-                domainType: TestCases.constants.domainType,
-                domainData: {
-                    name: await token.name(),
-                    version: data.domainVersion || '1',
-                    verifyingContract: tokenAddress,
-                    chainId: config.networks.hardhat.chainId,
-                },
-            });
+
+            console.log(owner.address, 'Owner address...');
+            // Handle tokenNonces for AAVE
+            // AAVE has nonce declared as _nonce get contract with abi
+            let tokenNonce;
+
+            if (data.symbol === 'AAVE') {
+                const aaveContract = await ethers.getContractAt(
+                    aaveABi,
+                    tokenAddress,
+                    owner
+                );
+
+                tokenNonce = parseInt(await aaveContract._nonces(owner.address));
+            } else {
+                tokenNonce = parseInt(await tokenNonces.nonces(owner.address));
+            }
+
+
+            console.log(tokenNonce, 'Nonce of token');
+            let approvalSigR, approvalSigS, approvalSigV;
+            // DAI
+            if (data.symbol === 'DAI') {
```

Comment:
1. use binaryPermitType variable to check which if else to use
2. abstract common code outside

Reply:
done